streams/shared/src/main/scala/zio/stream/ZStream.scala

```diff
@@ -1911,16 +1911,15 @@ class ZStream[-R, +E, +A] private[stream] (private[stream] val structure: ZStrea
         permits <- Semaphore.make(n.toLong).toManaged_
         _ <- self.foreachManaged { a =>
               for {
-                latch <- Promise.make[Nothing, Unit]
                 p     <- Promise.make[E1, B]
+                latch <- Promise.make[Nothing, Unit]
                 _     <- out.offer(Pull.fromPromise(p))
-                _     <- (permits.withPermit(latch.succeed(()) *> f(a).to(p))).fork
+                _     <- permits.withPermit(latch.succeed(()) *> f(a).to(p)).fork
```

Comment:
The parens were changed slightly here; should be:
```suggestion
                _     <- (permits.withPermit(latch.succeed(())) *> f(a).to(p)).fork
```
The purpose of the latch is to ensure that the forked fiber has started before moving on to the next element.

Reply:
Doesn't that change mean that we give back the permit as soon as we complete the latch? Wouldn't that result in higher than intended parallelism in the execution of the `f` effect?

Reply:
Yes, but the parallelism is limited by the capacity of the queue.

I actually think the permits are now redundant, but I'll need to look back to recall why they were added.

Reply:
Here is the original version:

```scala
(permits.withPermit(latch.succeed(()) *> f(a).to(p))).fork
```

It looks like there is one set of parentheses surrounding everything except `fork` so I think it is value preserving to remove them.

Reply:
I'm not sure if the queue is enough because without the semaphore we'll have all these fibers suspending to wait to offer the result of `f` to the queue but I think they will still all be doing `f` in parallel. I'll add a test for this and we can see if we can simplify.

Reply:
Ah you're right - the permits were surrounding everything. Good catch.

They won't all be suspending while executing `f`, as they first enqueue a `Pull.fromPromise`, and only then execute `f`. So once `n` have executed, the next one will be suspended on enqueueing the promise.

Reply:
Yes I think you're right the semaphore can be eliminated.